lang/java/idl/src/main/java/org/apache/avro/idl/Schemas.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.avro.idl;
+
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Field;
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.IdentityHashMap;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * Avro Schema utilities, to traverse...
+ */
+public final class Schemas {
+
+  private Schemas() {
+  }
+
+  /**
+   * Depth first visit.
+   */
+  public static <T> T visit(final Schema start, final SchemaVisitor<T> visitor) {
+    // Set of Visited Schemas
+    IdentityHashMap<Schema, Schema> visited = new IdentityHashMap<>();
+    // Stack that contains the Schemas to process and afterVisitNonTerminal
+    // functions.
+    // Deque<Either<Schema, Supplier<SchemaVisitorAction>>>
+    // Using Either<...> has a cost we want to avoid...
+    Deque<Object> dq = new ArrayDeque<>();
+    dq.push(start);
+    Object current;
+    while ((current = dq.poll()) != null) {
+      if (current instanceof Supplier) {
+        // We are executing a non-terminal post visit.
+        @SuppressWarnings("unchecked")
+        SchemaVisitorAction action = ((Supplier<SchemaVisitorAction>) current).get();
+        switch (action) {
+        case CONTINUE:
+          break;
+        case SKIP_SUBTREE:
+          throw new UnsupportedOperationException();
```

Comment:
add exception message ?!

Reply:
Yeah, I'll add one. This was copied from the `avro-compiler` module (there will be still 26 cases of this exception without a message left).